File: src/main/java/net/maple/handlers/user/UserMoveHandler.java
package net.maple.handlers.user;

import client.Character;
import client.Client;
import field.movement.MovePath;
import net.maple.SendOpcode;
import net.maple.handlers.PacketHandler;
import util.packet.Packet;
import util.packet.PacketReader;
import util.packet.PacketWriter;

public class UserMoveHandler extends PacketHandler {

    @Override
    public void handlePacket(PacketReader reader, Client c) {
        Character chr = c.getCharacter();

        reader.readLong(); // probably timestamp
        reader.read();
        reader.readLong();
        reader.readInteger();
        reader.readInteger();
        reader.readInteger();

        MovePath path = chr.move(reader);

        if (chr.getField() != null) {
            chr.getField().broadcast(movePlayer(chr, path), chr);
        }
    }

    private static Packet movePlayer(Character chr, MovePath path) {
        PacketWriter pw = new PacketWriter(32);

        pw.writeHeader(SendOpcode.USER_MOVE);
        pw.writeInt(chr.getId());
        path.encode(pw);

        return pw.createPacket();
    }
}

File: src/main/java/net/maple/handlers/user/UserMoveHandler.kt
package net.maple.handlers.user

import client.Client
import client.replay.MoveCollection.Movement
import constants.FieldConstants.JQ_FIELDS
import net.maple.handlers.PacketHandler
import net.maple.packets.CharacterPackets.move
import util.logging.LogType
import util.logging.Logger
import util.packet.PacketReader

class UserMoveHandler : PacketHandler {

    override fun handlePacket(reader: PacketReader, c: Client) {
        val chr = c.character

        reader.readInteger() // drInfo
        reader.readInteger() // ~v27
        val fieldKey = reader.readByte()
        if (fieldKey != chr.fieldKey) {
            return run {
                Logger.log(LogType.FIELD_KEY, "FieldKey mismatch", this, c)
                //c.close(this, "FieldKey mismatch")
            }
        }
        reader.readInteger() // ~v28
        reader.readInteger() // ~v29

        reader.readInteger() // crc
        reader.readInteger() // dwKey
        val crc32 = reader.readInteger() // crc32

        //println(crc32)

        val path = chr.move(reader)

        if (JQ_FIELDS.contains(chr.fieldId)) {
            chr.moveCollections[chr.fieldId]?.movements?.add(Movement(System.currentTimeMillis(), reader.data))
            chr.times["jq start"] = System.currentTimeMillis()
        }

        chr.field.broadcast(chr.move(path), chr)
    }
}